Matrix.dot can now calculate dot product between two complex vectors

amsuhane opened the pull request:
Changed the dot product function so that it can also calculate the Hermitian inner product

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fixes #14920  

#### Brief description of what is fixed or changed
Changed the dot product function so that it can now calculate Hermitian inner product.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
  * dot function can also calculate hermitian inner product.
<!-- END RELEASE NOTES -->

Commit:
Chnaged dot product function in matrices.py

Commit:
Corrected whitespace error

amsuhane commented:
@siefkenj are there any more changes to be made?

asmeurer commented on sympy/matrices/matrices.py:
```diff
@@ -2616,7 +2630,38 @@ def dot(self, b):
+
+        if(hermitian == True):
+            return mat.inner_product(b, conjugate_convention)
+        elif(hermitian == False):
```
Don't use parentheses with an if. You can also just write `if hermitian` or `elif not hermitian`. I recommend reading through [PEP 8](https://www.python.org/dev/peps/pep-0008/).

amsuhane replied:
I will go through PEP8 and correct the code.

asmeurer commented on sympy/matrices/matrices.py:
```diff
@@ -2616,7 +2630,38 @@ def dot(self, b):
+            warning_to_print =  "Warning: taking the non-Hermitian dot product of complex vectors. To suppress this warning, explicitly set hermitian=False."
+            if mat.is_symbolic() == True:
+                if b.is_symbolic() == False and b.conjugate() != b:
+                    print(warning_to_print)
```
Do we need warnings at all? It's generally better to not issue warnings unless absolutely necessary.

siefkenj replied:
I guess this should be consistent with the rest of sympy. I don't know if people who enter complex vectors into `dot` expect it to be the hermitian dot product or not...I think a warning here will save a lot of confusion from our users, but it isn't strictly necessary.

asmeurer commented on sympy/matrices/matrices.py:
```diff
@@ -2561,11 +2561,19 @@ def diagonal_solve(self, rhs):
+        of the two vectors. To calculate the inner product we call the inner_product function.
+
+        The expected kwargs are hermitian and conjugate_convention. Hermitian takes two True or False
+        depending on wether dot product is to be calculated or inner product is to be used.
```
Spelling

asmeurer commented on sympy/matrices/matrices.py:
```diff
@@ -2561,11 +2561,19 @@ def diagonal_solve(self, rhs):
+
+        The expected kwargs are hermitian and conjugate_convention. Hermitian takes two True or False
+        depending on wether dot product is to be calculated or inner product is to be used.
+        conjugate_convention has basically two inputs ("maths", "left", "math") or ("physics", "right").
```
This isn't clear that the input should be one of these. Also there should be an example showing how to use the argument.

asmeurer commented on sympy/matrices/matrices.py:
```diff
@@ -2616,7 +2630,38 @@ def dot(self, b):
             b = b.reshape(n, 1)
-        # Now ``mat`` is a row vector and ``b`` is a column vector.
+
+        """ Retrive the calue of hermitian and conjugate_convention from kwargs.
```
Typo

asmeurer replied:
Use `#` for comments. Multiline strings should only be used for docstrings.

asmeurer commented on sympy/matrices/matrices.py:
```diff
@@ -2835,6 +2880,26 @@ def gauss_jordan_solve(self, b, freevar=False):
+        elif conjugate_convention in ("physics", "right"):
+            return (mat*b.conjugate())[0]
+        else:
+            print("Invalid conjugate_convention is entered.")
```
This should raise ValueError, not print.

amsuhane commented on sympy/matrices/matrices.py:
```diff
@@ -2561,11 +2561,19 @@ def diagonal_solve(self, rhs):
+        The expected kwargs are hermitian and conjugate_convention. Hermitian takes two True or False
+        depending on wether dot product is to be calculated or inner product is to be used.
+        conjugate_convention has basically two inputs ("maths", "left", "math") or ("physics", "right").
+        conjugate_convention is used to  determine which convention is to be followed when calculating the inner product.
```
Ok, I will work on that.

amsuhane commented on sympy/matrices/matrices.py:
```diff
@@ -2579,10 +2587,16 @@ def dot(self, b):
         10
+        >>> q = Matrix([1j, 1j, 1j])
+        >>> q.dot(q, hermitian = False)
+        -3.00000000000000
```
Upon checking, it is a sympy number...

```
>>> from sympy import *
>>> a = Matrix([1j, 1j, 1j])
>>> a.dot(a)
-3.00000000000000
>>> type(a.dot(a))
<class 'sympy.core.numbers.Float'>

```

Commit:
Changes are as follows:
- Removed inner_product function.
- Moved functionality of inner_product function into dot product function
- Removed warning about unset hermitian flag
- Warning previously shown by print is now shown using Raise ValueError
- Corrected typos
- Minor changes to bring code in accordance to PEP 8 conventions

amsuhane commented:
I have resolved the above-reviewed issues. Are there any more changes to be made?

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -2579,10 +2598,21 @@ def dot(self, b):
         >>> v = [3, 2, 1]
         >>> M.row(0).dot(v)
         10
+        >>> q = Matrix([1j, 1j, 1j])
```
Do a `from sympy import I` and use the sympy imaginary unit for this example

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -2616,8 +2646,28 @@ def dot(self, b):
+        # is true but no conjugate_convention is not passed then
+        # automatically set it to "maths"
+
+        if (conjugate_convention != None) and (hermitian == None):
```
`is` and `is not` for `None` comparison

siefkenj reviewed (commented):
This is almost ready! Can you also add a couple of tests? Verify both conjugate conventions and explicitly setting `hermitian=False`.

Commit:
The changes are:
- added more tests for dot product function
- Few minor changes

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -2616,8 +2649,28 @@ def dot(self, b):
+            if conjugate_convention in ("maths", "left", "math"):
+                return (mat.conjugate()*b)[0]
+            elif conjugate_convention in ("physics", "right"):
+                return (mat*b.conjugate())[0]
```
I think this code would be better if you did `b=b.conjugate()` or `mat=mat.conjugate()` and then returned a value only once.

siefkenj commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -2503,6 +2503,12 @@ def test_dot():
+    assert Matrix([1, 2, 3*I]).dot(Matrix([1*I, 2, 3*I])) == -5 + 1*I
+    assert Matrix([1, 2, 3*I]).dot(Matrix([1*I, 2, 3*I]), hermitian = False) == -5 + 1*I
+    assert Matrix([1, 2, 3*I]).dot(Matrix([1*I, 2, 3*I]), hermitian = True) == 13 + 1*I
+    assert Matrix([1, 2, 3*I]).dot(Matrix([1*I, 2, 3*I]), hermitian = True, conjugate_convention = "physics") == 13.0 - 1.0*I
```
Why the floats?

amsuhane replied:
I have corrected it.

siefkenj commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -2503,6 +2503,12 @@ def test_dot():
     assert ones(1, 3).dot([1, 1, 1]) == 3
     assert Matrix([1, 2, 3]).dot(Matrix([1, 2, 3])) == 14
+    assert Matrix([1, 2, 3*I]).dot(Matrix([1*I, 2, 3*I])) == -5 + 1*I
+    assert Matrix([1, 2, 3*I]).dot(Matrix([1*I, 2, 3*I]), hermitian = False) == -5 + 1*I
```
kwargs have no spaces. e.g., `hermitian=False`, not `hermitian = False`

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -2616,8 +2649,28 @@ def dot(self, b):
+            elif conjugate_convention in ("physics", "right"):
+                return (mat*b.conjugate())[0]
+            else:
+                raise ValueError("Invalid conjugate_convention is entered.")
```
We can do better. "Unknown `conjugate_convention` '...'. `conjugate_convention` must be one of ..."

Commit:
Added minor changes to dot function and test for dot function

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,29 @@ def diagonal_solve(self, rhs):
+        arguments are: "maths", "left", "math", "physics" or "right".
+        "maths", "math" or "left" all correspond to the same convention
+        in which hermitian of the first vector is used.
+        i.e. (a.conjugate() * b)[0]
```
"in which the hermitian conjugate of the first vector is used."

The "i.e." part is a sentence fragment.  Should be "...is used, i.e., meh".

But also `a` is undefined, its not clear `*` DTRT here, I have no idea why `[0]` is there, and this needs to be inside dbl tics.  I do like the idea of having an example but this might do more harm than good: maybe just remove it?

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,29 @@ def diagonal_solve(self, rhs):
+        i.e. (a.conjugate() * b)[0]
+        "physics" or "right" both correspond to the same convention in
+        which hermitian of the second vector is used.
+        i.e. (a * b.conjugate())[0]
```
again "in which the hermitian conjugate of the second..."

And delete the "i.e." bit? :-(

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2616,7 +2649,28 @@ def dot(self, b):
+        if (conjugate_convention is not None) and (hermitian is None):
+            hermitian = True
+        if (hermitian == True) and (conjugate_convention is None):
+            conjugate_convention = "maths"
```
no parenthesis
`if hermitian and conjugate_conv is None:`

cbm755 reviewed (commented):
Sorry late to the party, I like this idea.  Here's a few suggestions from reading the changes.

Commit:
Minor changes

amsuhane commented:
Sorry for the mistakes, I have corrected them.

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,28 @@ def diagonal_solve(self, rhs):
+        whether dot product is to be calculated or hermitian inner
+        product is to be calculated.
+
+        Conjugate_convention takes one argument as input. The accepted
```
```suggestion
        If ``conjugate_convention`` is ``"left"``, ``"math"`` or ``"maths"``,
```

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,28 @@ def diagonal_solve(self, rhs):
+        product is to be calculated.
+
+        Conjugate_convention takes one argument as input. The accepted
+        arguments are: ``"maths"`` ,``"math"`` ,``"left"`` ,``"physics"``
```
```suggestion
        the conjugate of the first vector (``self``) is used.  If ``"right"``
```

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,28 @@ def diagonal_solve(self, rhs):
+
+        Conjugate_convention takes one argument as input. The accepted
+        arguments are: ``"maths"`` ,``"math"`` ,``"left"`` ,``"physics"``
+        or ``"right"``.
```
```suggestion
        or ``"physics"`` is specified, the conjugate of the second vector ``b`` is used.
```

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,28 @@ def diagonal_solve(self, rhs):
+        Conjugate_convention takes one argument as input. The accepted
+        arguments are: ``"maths"`` ,``"math"`` ,``"left"`` ,``"physics"``
+        or ``"right"``.
+        ``"maths"``, ``"math"`` or ``"left"`` all correspond to the same convention
```
```suggestion
```

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,28 @@ def diagonal_solve(self, rhs):
+        arguments are: ``"maths"`` ,``"math"`` ,``"left"`` ,``"physics"``
+        or ``"right"``.
+        ``"maths"``, ``"math"`` or ``"left"`` all correspond to the same convention
+        in which hermitian of the first vector is used.
```
```suggestion
```

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,28 @@ def diagonal_solve(self, rhs):
+        or ``"right"``.
+        ``"maths"``, ``"math"`` or ``"left"`` all correspond to the same convention
+        in which hermitian of the first vector is used.
+        ``"physics"`` or ``"right"`` both correspond to the same convention in
```
```suggestion
```

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,28 @@ def diagonal_solve(self, rhs):
+        ``"maths"``, ``"math"`` or ``"left"`` all correspond to the same convention
+        in which hermitian of the first vector is used.
+        ``"physics"`` or ``"right"`` both correspond to the same convention in
+        which hermitian of the second vector is used.
```
```suggestion
```

cbm755 commented on sympy/matrices/matrices.py:
```diff
@@ -2616,7 +2649,28 @@ def dot(self, b):
+
+        if conjugate_convention is not None and hermitian is None:
+            hermitian = True
+        if hermitian == True and conjugate_convention is None:
```
```suggestion
        if hermitian and conjugate_convention is None:
```

cbm755 commented:
I tried the Github beta feature of suggestions, let me know how it works.  Apparently this let's me suggest changes directly which you can then approve (or not).  Unfortunately, it seems to go line-by-line which isn't opitmal

cbm755 committed:
Minor change

Co-Authored-By: amsuhane <35529737+amsuhane@users.noreply.github.com>

cbm755 committed:
Minor change

Co-Authored-By: amsuhane <35529737+amsuhane@users.noreply.github.com>

cbm755 committed:
Minor change

Co-Authored-By: amsuhane <35529737+amsuhane@users.noreply.github.com>

cbm755 committed:
Minor change

Co-Authored-By: amsuhane <35529737+amsuhane@users.noreply.github.com>

cbm755 committed:
Minor change

Co-Authored-By: amsuhane <35529737+amsuhane@users.noreply.github.com>

cbm755 committed:
Minor change

Co-Authored-By: amsuhane <35529737+amsuhane@users.noreply.github.com>

cbm755 committed:
Minor change

Co-Authored-By: amsuhane <35529737+amsuhane@users.noreply.github.com>

cbm755 committed:
Minor change

Co-Authored-By: amsuhane <35529737+amsuhane@users.noreply.github.com>

Commit:
Minor changes

Commit:
Merge branch 'master' of https://github.com/amsuhane/sympy

amsuhane commented:
I tried using the Github Beta but the tests failed, so I made the changes my self.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,24 @@ def diagonal_solve(self, rhs):
+        must be either a matrix of size 1 x n, n x 1, or a list/tuple of length n.
+        A scalar is returned.
+
+        In case one or both of the vectors are complex and the hermitian
```
Change to:

```
By default, ``dot`` does not conjugate ``self`` or ``b``, even if there are 
complex entries. Set ``hermitian=True`` (and optionally a ``conjugate_convention``)
to compute the hermitian inner product.
```

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,24 @@ def diagonal_solve(self, rhs):
+        flag is set to ``True`` this will calculate the hermitian inner
+        product of the two vectors.
+
+        Possible kwargs are hermitian and conjugate_convention.
```
double backticks around kwargs here.

siefkenj commented on sympy/matrices/matrices.py:
```diff
@@ -2561,10 +2561,24 @@ def diagonal_solve(self, rhs):
+        product of the two vectors.
+
+        Possible kwargs are hermitian and conjugate_convention.
+        Hermitian takes argument as either True or False depending on
```
Remove this sentence.

siefkenj commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -2503,6 +2503,12 @@ def test_dot():
+    assert Matrix([1, 2, 3*I]).dot(Matrix([I, 2, 3*I]), hermitian=True) == 13 + I
+    assert Matrix([1, 2, 3*I]).dot(Matrix([I, 2, 3*I]), hermitian=True, conjugate_convention="physics") == 13 - I
+    assert Matrix([1, 2, 3*I]).dot(Matrix([4, 5*I, 6]), hermitian=True, conjugate_convention="right") == 4 + 8*I
+    assert Matrix([1, 2, 3*I]).dot(Matrix([4, 5*I, 6]), hermitian=True, conjugate_convention="left") == 4 - 8*I
```
Add a test that sets `conjugate_convention` without setting `hermitian` and a test that sets `hermitian=False` but sets `conjugate_convention="left"` and verifies that no conjugation occurs.

Commit:
Minor changes

siefkenj commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -2503,12 +2503,14 @@ def test_dot():
     assert Matrix([1, 2, 3*I]).dot(Matrix([4, 5*I, 6]), hermitian=True, conjugate_convention="right") == 4 + 8*I
     assert Matrix([1, 2, 3*I]).dot(Matrix([4, 5*I, 6]), hermitian=True, conjugate_convention="left") == 4 - 8*I
+    assert Matrix([I, 2*I, 3]).dot(Matrix([I, 2*I, 3]), hermitian=False, conjugate_convention="left") == 4
+    assert Matrix([I, 2*I, 3]).dot(Matrix([I, 2*I, 3]), conjugate_convention="left") == 14
```
These two tests would be easier to understand if the vector was `[I, 2*I]`

Commit:
Minor changes to test for dot function

siefkenj commented:
This looks to me like it's ready to merge! Good work @amsuhane

amsuhane commented:
Thank you :)

siefkenj commented:
@cbm755 @asmeurer Can one of you merge this?
